src/words/environment/WordsObject.java

```diff
@@ -13,13 +15,35 @@
 	private Position cell;
 	private String currentMessage;
 	private Action lastAction;
-
+	private HashMap<WordsObject, ArrayList<Property>> referers;
+	
+	public void updateReferers(WordsObject referringObject, Property referringProperty) {
+		ArrayList<Property> referringProperties;
+		if ((referringProperties = referers.get(referringObject)) == null) {
+			referringProperties = new ArrayList<Property>();
+			referers.put(referringObject, referringProperties);
+		}
+		
+		referringProperties.add(referringProperty);
+	}
+	
+	public void clearReferers() {
+		for (WordsObject referringObject : referers.keySet()) {
+			for (Property property : referers.get(referringObject)) {
+				property.type = PropertyType.NOTHING;
+				property.objProperty = null;
+			}
+			referers.remove(referringObject);
```

Comment:
Disagree here. I know what you mean about removing from a collection while iterating through it. But I am not iterating through the HashMap technically. I'm iterating through a copy of the keys to the HashMap.